src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -366,31 +414,98 @@ private void onSignal(
         private void onMqttConnect(
             MqttConnectFW packet)
         {
-            doMqttConnack();
+            int reasonCode = 0x00;
+
+            final String protocolName = packet.protocolName().asString();
+            if (!"MQTT".equals(protocolName) || packet.protocolVersion() != 5)
+            {
+                reasonCode = 0x84; // unsupported protocol version
+            }
+            else if ((packet.flags() & 0b0000_0001) != 0b0000_0000)
+            {
+                reasonCode = 0x81; // malformed packet
+            }
+
+            if (reasonCode == 0)
+            {
+                doMqttConnack(reasonCode);
+                this.decodeState = this::decodeSession;
+            }
+            else
+            {
+                doMqttConnack(reasonCode);
+                doEnd(decodeTraceId);
+            }
         }
 
         private void onMqttPingReq(
-            MqttPingReqFW packet)
+            MqttPingReqFW ping)
         {
             doMqttPingResp();
         }
 
-        private void onMqttSubscribe()
+        private void onMqttSubscribe(
+            MqttSubscribeFW subscribe)
         {
+            DirectBuffer buffer = subscribe.topicFilters().buffer();
+            int offset = subscribe.topicFilters().offset();
+            int length = subscribe.topicFilters().sizeof();
+            int topicsLength = 0;
+
+            if (length == 0)
+            {
+                doMqttDisconnect(0x82);
+                doEnd(decodeTraceId);
+            }
+            else
+            {
+                while (length > 0)
+                {
+                    MqttSubscriptionTopicFW topic = mqttSubscriptionTopicRO
+                        .tryWrap(buffer, offset, offset + length);
+
+                    offset += topic.sizeof();
+                    length -= topic.sizeof();
+
+                    topicsLength++;
+                }
+
+                doMqttSuback(topicsLength);
+            }
         }
 
-        private void onMqttUnsubscribe()
+        private void onMqttUnsubscribe(
+            MqttUnsubscribeFW unsubscribe)
         {
+            DirectBuffer buffer = unsubscribe.topicFilters().buffer();
+            int offset = unsubscribe.topicFilters().offset();
+            int length = unsubscribe.topicFilters().sizeof();
+
+            int topicsLength = 0;
+
+            while (length > 0)
+            {
+                MqttTopicFW topic = mqttTopicRO
+                    .tryWrap(buffer, offset, offset + length);
+
+                offset += topic.sizeof();
+                length -= topic.sizeof();
+
+                topicsLength++;
+            }
```

Comment:
Let's change this to be limit-based instead of length-based.

Suggest the following:
```
            final OctetsFW topicFilters = unsubscribe.topicFilters();
            final DirectBuffer buffer = topicFilters.buffer();
            final int limit = topicFilters.limit();
            final int offset = topicFilters.offset();

             int topics = 0;
             MqttTopicFW topic;
             for (int progress = offset; progress < limit; progress = topic.limit())
             {
                topic = mqttTopicRO.tryWrap(buffer, progress, limit);
                if (topic == null)
                {
                    break;
                }
                topics++;
             }
```